trie/sync.go

```diff
@@ -221,6 +223,16 @@ func (s *Sync) AddSubTrie(root common.Hash, path []byte, parent common.Hash, par
 	if rawdb.HasTrieNode(s.database, owner, inner, root, s.scheme) {
 		return
 	}
+	// If the node does not exist in the database or its content does not
+	// match the requested content, schedule it for retrieval.
+	//
+	// Notably, the existent node should be wiped first if path scheme is
+	// used, otherwise state will end up with a weird situation that parent
+	// node is inconsistent with children while they are all present in
+	// database.
+	if s.scheme == rawdb.PathScheme {
+		rawdb.DeleteTrieNode(s.database, owner, inner, common.Hash{} /* unused */, s.scheme)
```

Comment:
When we wind up here, we're being a bit wasteful. 
The `rawdb.HasTrieNode` already invoked `HasStorageTrieNode(db, owner, path, hash)`, and looked up the path. It found a node, however, the check `h.hash(data) == hash` failed, so it returned `false`. 
At this point, we do `DeleteTrieNode` -> `DeleteStorageTrieNode`, which adds a deletion to the batch -- regardless of whether there is anything there or not. ( deletion puts a tombstone in the database, so even if there's nothing there to begin with, a deletion is _not_ a no-op).


A more lenient use of resources would be something like, 
```golang
	ok, mismatch := rawdb.HasTrieNode(s.database, owner, path, root, s.scheme)
	if ok{
		return
	}
	// node is not present, however, mismatch == true means that _something_ is 
	// occupying that path (but does not match the hash). This can only occur in path-scheme
	if mismatch{
		rawdb.DeleteTrieNodeByPath(s.database, owner, path)
	}
```

Reply:
I guess my suggestion is moot, in the long run, since we're getting rid of the hash-check

Reply:
But: how will we implement this fix after getting rid of hash-checks?